File: src/models/ExpenseRequest.ts
/**
 * ExpenseRequest Model
 * Struktur data untuk klaim reimburse dalam aplikasi StableFlow
 * 
 * Equivalent to ExpenseRequest.java in Android Native:
 * public class ExpenseRequest { ... }
 */

export interface ExpenseRequest {
  id: string;
  userId: string;
  title: string;
  description: string;
  amount: number;
  currency: 'USDC' | 'USD' | 'IDR';
  category: ExpenseCategory;
  status: ExpenseStatus;
  receiptUrl?: string;
  attachments?: string[];
  submittedAt: string;
  processedAt?: string;
  approvedBy?: string;
  rejectionReason?: string;
  notes?: string;
}

// Expense with blockchain transaction data
export interface ExpenseWithTx extends ExpenseRequest {
  txSignature?: string;
  txExplorerUrl?: string;
  paidAt?: string;
  payerAddress?: string;
}

export type ExpenseCategory =
  | 'TRAVEL'
  | 'MEALS'
  | 'SUPPLIES'
  | 'EQUIPMENT'
  | 'SOFTWARE'
  | 'TRAINING'
  | 'ENTERTAINMENT'
  | 'UTILITIES'
  | 'OTHER';

export type ExpenseStatus =
  | 'PENDING'
  | 'UNDER_REVIEW'
  | 'APPROVED'
  | 'REJECTED'
  | 'PAID'
  | 'CANCELLED';

export const ExpenseCategoryLabels: Record<ExpenseCategory, string> = {
  TRAVEL: 'Perjalanan Dinas',
  MEALS: 'Makan & Minum',
  SUPPLIES: 'Perlengkapan Kantor',
  EQUIPMENT: 'Peralatan',
  SOFTWARE: 'Software & Lisensi',
  TRAINING: 'Pelatihan & Sertifikasi',
  ENTERTAINMENT: 'Hiburan Klien',
  UTILITIES: 'Utilitas',
  OTHER: 'Lainnya'
};

export const ExpenseStatusLabels: Record<ExpenseStatus, string> = {
  PENDING: 'Menunggu',
  UNDER_REVIEW: 'Dalam Review',
  APPROVED: 'Disetujui',
  REJECTED: 'Ditolak',
  PAID: 'Dibayar',
  CANCELLED: 'Dibatalkan'
};

export const ExpenseStatusColors: Record<ExpenseStatus, { bg: string; text: string }> = {
  PENDING: { bg: 'bg-yellow-500/20', text: 'text-yellow-400' },
  UNDER_REVIEW: { bg: 'bg-blue-500/20', text: 'text-blue-400' },
  APPROVED: { bg: 'bg-green-500/20', text: 'text-green-400' },
  REJECTED: { bg: 'bg-red-500/20', text: 'text-red-400' },
  PAID: { bg: 'bg-emerald-500/20', text: 'text-emerald-400' },
  CANCELLED: { bg: 'bg-gray-500/20', text: 'text-gray-400' }
};

// Factory function to create a new ExpenseRequest
export const createExpenseRequest = (
  userId: string,
  title: string,
  description: string,
  amount: number,
  category: ExpenseCategory,
  receiptUrl?: string
): ExpenseRequest => {
  return {
    id: generateExpenseId(),
    userId,
    title,
    description,
    amount,
    currency: 'USDC',
    category,
    status: 'PENDING',
    receiptUrl,
    submittedAt: new Date().toISOString(),
  };
};

// Generate unique expense ID
const generateExpenseId = (): string => {
  const timestamp = Date.now().toString(36);
  const randomStr = Math.random().toString(36).substring(2, 8);
  return `EXP-${timestamp}-${randomStr}`.toUpperCase();
};

// Validation function
export const validateExpenseRequest = (expense: Partial<ExpenseRequest>): { valid: boolean; errors: string[] } => {
  const errors: string[] = [];

  if (!expense.title || expense.title.trim().length < 3) {
    errors.push('Judul harus minimal 3 karakter');
  }

  if (!expense.description || expense.description.trim().length < 10) {
    errors.push('Deskripsi harus minimal 10 karakter');
  }

  if (!expense.amount || expense.amount <= 0) {
    errors.push('Jumlah harus lebih dari 0');
  }

  if (expense.amount && expense.amount > 100000) {
    errors.push('Jumlah maksimal 100,000 USDC');
  }

  if (!expense.category) {
    errors.push('Kategori harus dipilih');
  }

  return {
    valid: errors.length === 0,
    errors
  };
};

File: android/app/src/main/java/com/stableflow/app/data/model/ExpenseRequest.java
package com.stableflow.app.data.model;

import java.io.Serializable;

/**
 * ExpenseRequest Model
 * Represents an expense claim submitted by an employee
 */
public class ExpenseRequest implements Serializable {

    private String id;
    private String userId;
    private String title;
    private String description;
    private double amount;
    private String currency;
    private String category;
    private String status;
    private String receiptUrl;
    private String submittedAt;
    private String processedAt;
    private String approvedBy;
    private String rejectionReason;
    private String notes;
    private String txSignature;
    private String txExplorerUrl;
    private String paidAt;
    private String payerAddress;
    private Double latitude;
    private Double longitude;
    private String locationAddress;

    // Default constructor for Firebase
    public ExpenseRequest() {
    }

    // Constructor with required fields
    public ExpenseRequest(String userId, String title, String description,
            double amount, String category) {
        this.userId = userId;
        this.title = title;
        this.description = description;
        this.amount = amount;
        this.currency = "USDC";
        this.category = category;
        this.status = "PENDING";
        this.submittedAt = java.time.Instant.now().toString();
    }

    // Getters and Setters
    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getUserId() {
        return userId;
    }

    public void setUserId(String userId) {
        this.userId = userId;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public double getAmount() {
        return amount;
    }

    public void setAmount(double amount) {
        this.amount = amount;
    }

    public String getCurrency() {
        return currency;
    }

    public void setCurrency(String currency) {
        this.currency = currency;
    }

    public String getCategory() {
        return category;
    }

    public void setCategory(String category) {
        this.category = category;
    }

    public String getStatus() {
        return status;
    }

    public void setStatus(String status) {
        this.status = status;
    }

    public String getReceiptUrl() {
        return receiptUrl;
    }

    public void setReceiptUrl(String receiptUrl) {
        this.receiptUrl = receiptUrl;
    }

    public String getSubmittedAt() {
        return submittedAt;
    }

    public void setSubmittedAt(String submittedAt) {
        this.submittedAt = submittedAt;
    }

    public String getProcessedAt() {
        return processedAt;
    }

    public void setProcessedAt(String processedAt) {
        this.processedAt = processedAt;
    }

    public String getApprovedBy() {
        return approvedBy;
    }

    public void setApprovedBy(String approvedBy) {
        this.approvedBy = approvedBy;
    }

    public String getRejectionReason() {
        return rejectionReason;
    }

    public void setRejectionReason(String rejectionReason) {
        this.rejectionReason = rejectionReason;
    }

    public String getNotes() {
        return notes;
    }

    public void setNotes(String notes) {
        this.notes = notes;
    }

    public String getTxSignature() {
        return txSignature;
    }

    public void setTxSignature(String txSignature) {
        this.txSignature = txSignature;
    }

    public String getTxExplorerUrl() {
        return txExplorerUrl;
    }

    public void setTxExplorerUrl(String txExplorerUrl) {
        this.txExplorerUrl = txExplorerUrl;
    }

    public String getPaidAt() {
        return paidAt;
    }

    public void setPaidAt(String paidAt) {
        this.paidAt = paidAt;
    }

    public String getPayerAddress() {
        return payerAddress;
    }

    public void setPayerAddress(String payerAddress) {
        this.payerAddress = payerAddress;
    }

    public Double getLatitude() {
        return latitude;
    }

    public void setLatitude(Double latitude) {
        this.latitude = latitude;
    }

    public Double getLongitude() {
        return longitude;
    }

    public void setLongitude(Double longitude) {
        this.longitude = longitude;
    }

    public String getLocationAddress() {
        return locationAddress;
    }

    public void setLocationAddress(String locationAddress) {
        this.locationAddress = locationAddress;
    }

    // Helper methods
    public boolean isPending() {
        return "PENDING".equals(status) || "UNDER_REVIEW".equals(status);
    }

    public boolean isApproved() {
        return "APPROVED".equals(status);
    }

    public boolean isPaid() {
        return "PAID".equals(status);
    }

    public boolean isRejected() {
        return "REJECTED".equals(status);
    }

    public String getStatusLabel() {
        switch (status) {
            case "PENDING":
                return "Menunggu";
            case "UNDER_REVIEW":
                return "Dalam Review";
            case "APPROVED":
                return "Disetujui";
            case "REJECTED":
                return "Ditolak";
            case "PAID":
                return "Dibayar";
            case "CANCELLED":
                return "Dibatalkan";
            default:
                return status;
        }
    }

    public String getCategoryLabel() {
        switch (category) {
            case "TRAVEL":
                return "Perjalanan Dinas";
            case "MEALS":
                return "Makan & Minum";
            case "SUPPLIES":
                return "Perlengkapan Kantor";
            case "EQUIPMENT":
                return "Peralatan";
            case "SOFTWARE":
                return "Software & Lisensi";
            case "TRAINING":
                return "Pelatihan";
            case "ENTERTAINMENT":
                return "Hiburan Klien";
            case "UTILITIES":
                return "Utilitas";
            case "OTHER":
                return "Lainnya";
            default:
                return category;
        }
    }
}
